vault/src/vaults.rs

```diff
@@ -99,44 +122,69 @@ impl<P: RelayPallet + BtcRelayPallet, B: BitcoinCoreApi + Clone> VaultTheftMonit
         Ok(())
     }
 
-    async fn check_transaction(
+    async fn stream_relayed_transactions(
         &self,
-        tx: Transaction,
-        block_hash: BlockHash,
+        rpc: B,
+        from_height: u32,
         num_confirmations: u32,
-    ) -> Result<(), Error> {
-        let tx_id = tx.txid();
-        tracing::debug!("Checking transaction: {}", tx_id);
-
-        // at this point we know that the transaction has `num_confirmations` on the bitcoin chain,
-        // but the relay can introduce a delay, so wait until the relay also confirms the transaction.
-        self.btc_parachain
-            .wait_for_block_in_relay(H256Le::from_bytes_le(&block_hash.to_vec()), Some(num_confirmations))
-            .await?;
+    ) -> impl Stream<Item = Result<(BlockHash, Transaction), Error>> + Send + Unpin + '_ {
+        Box::pin(
+            stream_blocks(rpc, from_height, num_confirmations)
+                .await
+                .then(move |block| async move {
+                    let transactions: Box<dyn Stream<Item = _> + Unpin + Send> = match block {
+                        Ok(e) => {
+                            let block_hash = e.block_hash();
+                            // at this point we know that the transaction has `num_confirmations` on the bitcoin chain,
+                            // but the relay can introduce a delay, so wait until the relay also confirms the
+                            // transaction.
+                            let transactions: Box<dyn Stream<Item = _> + Unpin + Send> = match self
+                                .btc_parachain
+                                .wait_for_block_in_relay(
```

Comment:
Yea I guess it might be possible. I think we need to review the entire code for correct fork handling, I think I saw some todos that mentioned that forks might not be handled correctly. Let's make a separate ticket for that?

Reply:
Yeah can you make a ticket for this?

Reply:
[ticket](https://www.notion.so/interlay/Check-correct-bitcoin-fork-handling-in-vault-client-d854a04ddd25454ab4aa165e16aca732)